File: api/app/schemas/applications.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

STATUS_VALUES = Literal[
    "saved", "tailoring", "applied", "oa_screen",
    "interview", "offer", "rejected", "withdrawn"
]

STATUS_LABELS = {
    "saved": "Saved",
    "tailoring": "Tailoring",
    "applied": "Applied",
    "oa_screen": "OA / Screen",
    "interview": "Interview",
    "offer": "Offer",
    "rejected": "Rejected",
    "withdrawn": "Withdrawn",
}


class ApplicationCreate(BaseModel):
    job_id: uuid.UUID
    resume_variant_id: uuid.UUID | None = None
    status: STATUS_VALUES = "saved"
    notes: str | None = None


class ApplicationUpdate(BaseModel):
    status: STATUS_VALUES | None = None
    resume_variant_id: uuid.UUID | None = None
    applied_at: datetime | None = None
    referral_contact: str | None = None
    notes: str | None = None
    follow_up_date: date | None = None


class ApplicationOut(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    company: str
    role_title: str
    jd_text: str | None
    jd_url: str | None
    status: str
    applied_at: datetime | None
    referral_contact: str | None
    notes: str | None
    follow_up_date: date | None
    ats_score: int | None
    variant_id: uuid.UUID | None
    variant_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ApplicationStats(BaseModel):
    total: int
    by_status: dict[str, int]

File: frontend/src/api/applications.ts
import { apiFetch } from "./client";

export type AppStatus =
  | "saved" | "tailoring" | "applied" | "oa_screen"
  | "interview" | "offer" | "rejected" | "withdrawn";

export const STATUS_LABELS: Record<AppStatus, string> = {
  saved: "Saved",
  tailoring: "Tailoring",
  applied: "Applied",
  oa_screen: "OA / Screen",
  interview: "Interview",
  offer: "Offer",
  rejected: "Rejected",
  withdrawn: "Withdrawn",
};

export const STATUS_ORDER: AppStatus[] = [
  "saved", "tailoring", "applied", "oa_screen",
  "interview", "offer", "rejected", "withdrawn",
];

export interface Application {
  id: string;
  job_id: string;
  company: string;
  role_title: string;
  jd_url: string | null;
  status: AppStatus;
  applied_at: string | null;
  referral_contact: string | null;
  notes: string | null;
  follow_up_date: string | null;
  ats_score: number | null;
  variant_id: string | null;
  variant_count: number;
  created_at: string;
  updated_at: string;
}

export interface ApplicationStats {
  total: number;
  by_status: Record<string, number>;
}

export const listApplications = () => apiFetch<Application[]>("/applications");
export const getApplication = (id: string) => apiFetch<Application>(`/applications/${id}`);
export const getStats = () => apiFetch<ApplicationStats>("/applications/stats");

export const createApplication = (job_id: string) =>
  apiFetch<Application>("/applications", {
    method: "POST",
    body: JSON.stringify({ job_id }),
  });

export const updateApplication = (
  id: string,
  data: Partial<{
    status: AppStatus;
    notes: string;
    referral_contact: string;
    follow_up_date: string;
    applied_at: string;
  }>
) =>
  apiFetch<Application>(`/applications/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });

export const deleteApplication = (id: string) =>
  apiFetch<void>(`/applications/${id}`, { method: "DELETE" });
